### PR TITLE
Add pipelineRef remote resolution

### PR DESCRIPTION
**Pull request by sbwsg**
# Changes

Prior to this commit the API fields for remote resolution were just
stubs with no functionality.

This commit adds initial functionality: support in `pipelineRef`. You can now pull in pipelines directly from a git repo. This feature is in alpha. To actually see anything happening a dev also needs to deploy [`tektoncd/resolution`](https://github.com/tektoncd/resolution) and the [`gitresolver`](https://github.com/tektoncd/resolution/tree/main/gitresolver).

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
Initial integration with tektoncd/resolution project, allowing pipelineRefs to be resolved directly from public git repos.
```

### CODE REVIEW

**Comment by sbwsg** [comment]
/hold

Holding for more documentation + more reconciler tests + better coverage numbers.

**Comment by sbwsg** [comment]
Adding a short copy/pastable PipelineRun that can be used for manual testing:

```yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: reqpr
spec:
  params:
  - name: name
    value: Scott
  pipelineRef:
    resolver: git
    resource:
    - name: url
      value: https://github.com/sbwsg/catalog.git
    - name: commit
      value: f0629ad5f9046bb8ff853f5303a5158e6dba409c
    - name: path
      value: /pipeline/simple/0.1/simple.yaml
```

File: pkg/reconciler/pipelinerun/resources/pipelineref_test.go
Comment (lbernick) [inline pkg/reconciler/pipelinerun/resources/pipelineref_test.go:296]: might also be worth adding a test case where the bytes returned are invalid

File: pkg/reconciler/pipelinerun/controller.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/controller.go:93]: Yes you're absolutely right, and this is why the integration tests are all failing badly.

The only workaround I can think of right now is to put this initialization behind an environment variable that the user must explicitly enable.

Comment (sbwsg) [reply pkg/reconciler/pipelinerun/controller.go:93]: OK, unfortunately it doesn't look like I can optionally start informers like that, I'll have to think of something else.

**Comment by sbwsg** [comment]
Alpha integration failed first on `workspace-in-sidecar` and this time on `run-steps-as-non-root`. The failures don't appear to be directly related to the content of the pull request.

**Comment by sbwsg** [comment]
`workspace-in-sidecar` failed again.

/test pull-tekton-pipeline-alpha-integration-tests

File: pkg/reconciler/pipelinerun/controller.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/controller.go:93]: I saw the slack conversation you opened on pipeline-dev; I agree with Matt Moor that it would be OK to import the ResourceRequest CRD as part of the Tekton pipelines config.

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: I think what is specifically confusing is that the remote resolution is returning an error, but the pipelinerun resolution is not returning an error. For taskruns contained in pipelineruns, if a taskrun is still running, we treat this as status "pending", not an error. It would make more sense to me to handle a ref that is still resolving the same way. would it be possible to treat the ResourceRequest as "pending" rather than an error?

Comment (sbwsg) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: Ah I see, so if we updated the signature of `resources.GetPipelineData` to something like this:

```go
pipelineMeta, pipelineSpec, pending, err := resources.GetPipelineData(ctx, pr, getPipelineFunc)
switch {
case pending:
  // The code here to mark the PR as pending
case err != nil:
  // The code here to fail the PipelineRun
default:
  // The code to store the pipeline meta and spec on pr.status
}
```

Does that read clearer than the sentinel check we have here atm?

**Comment by sbwsg** [comment]
`TestExamples/v1beta1/taskruns/alpha/workspace-in-sidecar 1m2s`

See #4169 for progress on that test.

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: After discussing offline and spending a bit of time trying to rejig this bit of the code I think I'm going to leave it as-is for the time being.

There's scope here for a tidier interface between the reconcilers and pipelineref resolution but the codebase is in a bit of an in-between state because we have resolution logic in (a) the reconcilers themselves (b) the resources subpackage which abstracts some resolution mechanics and (c) the remote package which hides away the built-in OCI bundles resolution (and now the new remoteresolution functionality).

A holistic change to the way this all works makes sense, and is def something we can look at in another pr, but for this one I'm trying to avoid changing how tekton's existing resolution mechanics function to limit blast radius of the new code.

File: pkg/reconciler/pipelinerun/controller.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/controller.go:93]: I've brought the CRD over into pipelines. One unfortunate downside is that this means the CRD is duplicated both in resolution and here. I tried vendoring the CRD yaml in resolution but `go mod` doesn't pull anything except go code into pipelines.

File: pkg/apis/pipeline/v1beta1/taskref_validation.go
Comment (sbwsg) [reply pkg/apis/pipeline/v1beta1/taskref_validation.go:74]: I've had a go at tackling these readability nits and added a unit test, should be ready for another look!

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:8786]: Good point, this isn't something I've tackled as part of this PR but I am also starting to think about how the `resolution` package could provide test helpers / mocks for this so that pipelines doesn't have to own them.

File: pkg/reconciler/pipelinerun/resources/pipelineref.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/resources/pipelineref.go:129]: Added a comment.

**Comment by lbernick** [comment]
> Thanks for the review @lbernick I think it's ready for another look.
Sounds good, I can review tomorrow morning

File: pkg/reconciler/pipelinerun/controller.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/controller.go:93]: ack -- that's too bad but it doesn't sound like there's a good way around this duplication.

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: SGTM! Would you be able to create an issue or something to describe a refactoring that would make more sense?

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:8786]: consider adding a TODO comment here to move out of this test file

**Review by lbernick — COMMENTED** [review]
/lgtm

thanks Scott!

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: We do have an issue around this, #3305, so I've added a comment describing the remaining work to do here: https://github.com/tektoncd/pipeline/issues/3305#issuecomment-1061755446

Comment (lbernick) [reply pkg/reconciler/pipelinerun/pipelinerun.go:343]: sounds good, thanks!

File: pkg/reconciler/pipelinerun/pipelinerun_test.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/pipelinerun_test.go:8786]: I created an issue in `resolution` to capture this: https://github.com/tektoncd/resolution/issues/10

**Comment by sbwsg** [comment]
I've renamed the CRD from `ResourceRequest` to `ResolutionRequest`. All other feedback so far is addressed and the PR is ready for other reviewers to take a look.

File: config/300-resolutionrequest.yaml
Comment (lbernick) [reply config/300-resolutionrequest.yaml:33]: ah got it! thanks for the clarification

Comment (sbwsg) [reply config/300-resolutionrequest.yaml:38]: Added a comment, apologies for the lgtm churn!

**Comment by jerop** [comment]
@sbwsg given the hold, may I move this to the next milestone? hoping to release 0.34 today

File: pkg/reconciler/pipelinerun/controller.go
Comment (sbwsg) [reply pkg/reconciler/pipelinerun/controller.go]: I've removed the CRD from resolution. Now the definition in Pipelines is the only one and Tekton Resolution will lean on that rather than the other way round.

**Comment by sbwsg** [comment]
workspace-in-sidecar and another example failed. neither related to the PR.

/test pull-tekton-pipeline-alpha-integration-tests

**Comment by sbwsg** [comment]
@tektoncd/core-maintainers Feedback from Pipelines WG has been addressed and this PR is ready for another look. Cheers!

**Comment by sbwsg** [comment]
Rebased on main.

**Review by dibyom — APPROVED** [review]

File: config/300-resolutionrequest.yaml
Comment (sbwsg) [reply config/300-resolutionrequest.yaml:21]: Ah, good question  :S

I think it's used to filter which resources to install when a user executes `kubectl apply`. This isn't used by any of the other pipeline crds so I'll remove it. Cheers!

**Comment by sbwsg** [comment]
Updated resolution to be based on latest knative + go1.17.

Should be ready for one final look + (hopefully 🤞) lgtm!

**Comment by sbwsg** [comment]
Also we have a first draft of a Bundles resolver: https://github.com/tektoncd/resolution/tree/main/bundleresolver

**Comment by sbwsg** [comment]
Workspace-in-sidecar failed; known flake

/test pull-tekton-pipeline-alpha-integration-tests

**Comment by sbwsg** [comment]
TestPipelineTaskTimeout failed with

```
reason: TaskRunTimeout
status: "False"
type: Succeeded
```

I can't reproduce this locally so let's see if it repeats.

/test pull-tekton-pipeline-alpha-integration-tests

**Pull request merged**
